src/lm/arpa-sampling.cc

```diff
@@ -0,0 +1,169 @@
+  if (it != probs_[order - 1].end()) {
+    WordToProbsMap::const_iterator it2 = probs_[order - 1][history].find(word);
+    if (it2 != probs_[order - 1][history].end()) {
+      bow = it2->second.second;
+    }
+  }
+  return bow;
+}
+
+void ArpaSampling::GetUnigramDistribution(std::vector<BaseFloat> *unigram_probs) {
+  for (int32 i = 0; i < num_words_; i++) {
+    HistType h; // empty history
+    WordToProbsMap::const_iterator it = probs_[0][h].find(i);
+    if (it != probs_[0][h].end()) {
+      (*unigram_probs)[i] = exp(it->second.first);
+    }    
+  }
+}
+ 
+void ArpaSampling::ComputeHistoriesWeights(
+    const std::vector<std::pair<HistType, BaseFloat> > &histories,
+    HistWeightsType *hists_weights) {
+  for (std::vector<std::pair<HistType, BaseFloat> >::const_iterator
+      it = histories.begin(); it != histories.end(); ++it) {
+    HistType history((*it).first);
+    KALDI_ASSERT(history.size() <= ngram_order_);
+    for (int32 i = 0; i < history.size() + 1; i++) {
+      HistType h_tmp = history;
+      // (*it).second is the input weight of a history 
+      BaseFloat prob = 1.0 / histories.size() * ((*it).second);
+      while (h_tmp.size() > (history.size() - i)) {
+        HistType::iterator last = h_tmp.end() - 1;
+        HistType h(h_tmp.begin(), last);
+        int32 word = h_tmp.back();
+        prob *= exp(GetBackoffWeight(h_tmp.size(), word, h));
+        HistType h_up(h_tmp.begin() + 1, h_tmp.end());
+        h_tmp = h_up;
+      }
+      HistType::iterator begin = history.begin() + i;
+      HistType h(begin, history.end());
+      (*hists_weights)[h] += prob;
+    }
+  }
+}
+
+BaseFloat ArpaSampling::GetOutputWordsAndAlpha(const std::vector<std::pair<HistType, BaseFloat> > &histories,
+    unordered_map<int32, BaseFloat> *pdf_w) {
+  HistWeightsType hists_weights;
+  ComputeHistoriesWeights(histories, &hists_weights); 
+  BaseFloat prob = 0;
+  for (HistWeightsType::const_iterator it = hists_weights.begin(); it != hists_weights.end(); ++it) {
+    HistType h(it->first);
+    int32 order = h.size();
+    NgramType::const_iterator it_hist = probs_[order].find(h);
+    if (it_hist != probs_[order].end()) {
+      for(WordToProbsMap::const_iterator it_word = probs_[order][h].begin(); 
+          it_word != probs_[order][h].end(); ++it_word) {
+        int32 word = it_word->first;
+        if (order > 0) {
+          HistType::iterator last = h.end() - 1;
+          HistType::iterator first = h.begin() + 1;
+          HistType h1(h.begin(), last);
+          HistType h2(first, h.end());
+          prob = it->second * (exp(probs_[order][h][word].first) - 
+                  exp(GetBackoffWeight(order, h.back(), h1) + GetProb(order, word, h2)));
+          unordered_map<int32, BaseFloat>::iterator map_it = (*pdf_w).find(word);
+          if (map_it != (*pdf_w).end()) {
+            (*pdf_w)[word] += prob;
+          } else {
+            (*pdf_w).insert({word, prob});
+          }
```

Comment:
`std::unordered_map::operator[]` [specification says](http://en.cppreference.com/w/cpp/container/unordered_map/operator_at): 
> Returns a reference to the value that is mapped to a key equivalent to *key*, performing an insertion if such key does not already exist [...] When the default allocator is used, this results in the key being copy constructed from *key* and the mapped value being value-initialized.  

So if `word` in not in the map, applying the `[]` inserts a new entry for that word with the value initialized to 0, the  (since the default allocator is used). Therefore, there is no need to call `find` to check whether an element in the map or not; just writing
```
(*pdf_w)[word] += prob;  // 'word' will be added if not found in map.
```
is sufficient. However, a comment on using the explicit but not extremely obvious behavior is helpful, or rather even necessary!